src/Sound/SoundHandler.cpp

```diff
@@ -0,0 +1,182 @@
+/*****************************************************************************
+ * Copyright (c) 2018-2020 openblack developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/openblack/openblack
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#include "SoundHandler.h"
+
+#include "3D/Camera.h"
+#include <Game.h>
+#include <fstream>
+#include <vector>
+
+namespace openblack::audio
+{
+
+SoundHandler::SoundHandler(std::unique_ptr<AudioDecoder>&& decoder, std::unique_ptr<AudioPlayer>&& player) : _audioDecoder(std::move(decoder)), _audioPlayer(std::move(player))
+{
+	_emitterId = 0;
+}
+
+void SoundHandler::RegisterSoundPack(std::unique_ptr<SoundPack>& soundPack)
+{
+	spdlog::debug("Registering SoundPack {}", soundPack->GetFileName());
+	_soundPacks.push_back(std::move(soundPack));
+	auto last = _soundPacks.back();
+	for (auto& [id, sound] : last->GetSounds())
+	{
+		if (_soundIdLookup.find(id) == _soundIdLookup.end())
+			_soundIdLookup[id] = last;
+		else
+			spdlog::warn("Duplicate sound ID (\"{}\"). Unable to add sound", id);
+	}
+}
+
+void SoundHandler::Tick(Game& game)
+{
+	// Update listener state
+	auto& cam = game.GetCamera();
+	auto pos = cam.GetPosition();
+	auto vel = cam.GetVelocity();
+	auto forward = cam.GetForward();
+	auto top = cam.GetUp();
+	_audioPlayer->UpdateListenerState(pos, vel, forward, top);
+
+	// Update active user emitter states
+	for (auto& [emitterId, emitter] : _emitters) _audioPlayer->UpdateEmitterState(emitter);
+
+	// Sort activated emitters
+	// TODO: Implement
+
+	// Release hardware resources
+	// Update emitters with hardware resources
+	// Assign hardware resources
+//	for (auto& [id, emitter] : _emitters)
+//	{
+//		_audioPlayer->CleanUpResource(emitter);
+//	}
+
+	// Cull stopped emitters
+	for (auto iterator = _emitters.begin(); iterator != _emitters.end();)
+	{
+		auto& [id, emitter] = *iterator;
+		auto audioStatus = _audioPlayer->GetAudioStatus(emitter.audioSourceId);
+
+		if (emitter.loop == PlayType::Once && audioStatus == AudioStatus::Stopped)
+		{
+			_audioPlayer->CleanUpResource(emitter);
+			iterator = _emitters.erase(iterator);
+		}
+		else
+			iterator++;
+	}
+}
+
+std::unique_ptr<Sound>& SoundHandler::GetSound(SoundId id)
+{
+	auto& sound = _soundIdLookup[id]->GetSound(id);
```

Comment:
I get a crash here during runtime because the map is accessed with `operator[]` without any checks for if id is within the map.

Reply:
I'm going to throw a custom std::runtime_error called `SoundNotFoundError` here. GetSound is called in a lot of different places and checks should be done during game startup to ensure we have the sounds.